File: java/src/beecrypt/security/DSAPublicKeyImpl.java
package beecrypt.security;

import java.io.*;
import java.math.BigInteger;
import java.security.*;
import java.security.interfaces.*;
import java.security.spec.*;

import beecrypt.io.*;

public class DSAPublicKeyImpl implements DSAPublicKey, Cloneable {
	private DSAParameterSpec _params;
	private BigInteger _y;
	private byte[] _enc;

	public DSAPublicKeyImpl(BigInteger p, BigInteger q, BigInteger g,
			BigInteger y) {
		_params = new DSAParameterSpec(p, q, g);
		_y = y;
	}

	public DSAPublicKeyImpl(DSAParameterSpec params, BigInteger y) {
		_params = params;
		_y = y;
	}

	public DSAPublicKeyImpl(DSAPublicKeySpec spec) {
		_params = new DSAParameterSpec(spec.getP(), spec.getQ(), spec.getG());
		_y = spec.getY();
	}

	public DSAPublicKeyImpl(DSAPublicKey key) {
		_params = new DSAParameterSpec(key.getParams().getP(), key.getParams()
				.getQ(), key.getParams().getG());
		_y = key.getY();
	}

	public Object clone() {
		return new DSAPublicKeyImpl(this);
	}

	public boolean equals(Object obj) {
		if (this == obj)
			return true;

		if (obj instanceof DSAPublicKey) {
			DSAPublicKey pri = (DSAPublicKey) obj;

			if (!pri.getParams().getP().equals(_params.getP()))
				return false;
			if (!pri.getParams().getQ().equals(_params.getQ()))
				return false;
			if (!pri.getParams().getG().equals(_params.getG()))
				return false;
			if (!pri.getY().equals(_y))
				return false;

			return true;
		}
		return false;
	}

	public DSAParams getParams() {
		return _params;
	}

	public BigInteger getY() {
		return _y;
	}

	public byte[] getEncoded() {
		if (_enc == null) {
			try {
				ByteArrayOutputStream bos = new ByteArrayOutputStream();
				BeeOutputStream bee = new BeeOutputStream(bos);

				bee.writeBigInteger(_params.getP());
				bee.writeBigInteger(_params.getQ());
				bee.writeBigInteger(_params.getG());
				bee.writeBigInteger(_y);
				bee.close();

				_enc = bos.toByteArray();
			} catch (IOException e) {
				// shouldn't occur
			}
		}

		return _enc;
	}

	public String getAlgorithm() {
		return "DSA";
	}

	public String getFormat() {
		return "BEE";
	}
}

File: c++/provider/DSAPublicKeyImpl.cxx
/*
 * Copyright (c) 2004 Beeyond Software Holding BV
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include "beecrypt/c++/provider/DSAPublicKeyImpl.h"
#include "beecrypt/c++/io/ByteArrayOutputStream.h"
using beecrypt::io::ByteArrayOutputStream;
#include "beecrypt/c++/beeyond/BeeOutputStream.h"
using beecrypt::beeyond::BeeOutputStream;

using namespace beecrypt::provider;

namespace {
    const String FORMAT_BEE("BEE");
    const String ALGORITHM_DSA("DSA");
}

DSAPublicKeyImpl::DSAPublicKeyImpl(const DSAPublicKey& copy)
{
	_params = new DSAParameterSpec(copy.getParams());
	_y = copy.getY();
	_enc = 0;
}

DSAPublicKeyImpl::DSAPublicKeyImpl(const DSAPublicKeyImpl& copy) : _y(copy._y)
{
	_params = new DSAParameterSpec(*copy._params);
	_enc = 0;
}

DSAPublicKeyImpl::DSAPublicKeyImpl(const DSAParams& params, const BigInteger& y) : _y(y)
{
	_params = new DSAParameterSpec(params.getP(), params.getQ(), params.getG());
	_enc = 0;
}

DSAPublicKeyImpl::DSAPublicKeyImpl(const dsaparam& params, const mpnumber& y) : _y(y)
{
	_params = new DSAParameterSpec(BigInteger(params.p), BigInteger(params.q), BigInteger(params.g));
	_enc = 0;
}

DSAPublicKeyImpl::DSAPublicKeyImpl(const BigInteger& p, const BigInteger& q, const BigInteger& g, const BigInteger& y)
{
	_params = new DSAParameterSpec(p, q, g);
	_y = y;
	_enc = 0;
}

DSAPublicKeyImpl::~DSAPublicKeyImpl()
{
	delete _params;
	delete _enc;
}

DSAPublicKeyImpl* DSAPublicKeyImpl::clone() const throw ()
{
	return new DSAPublicKeyImpl(*this);
}

bool DSAPublicKeyImpl::equals(const Object* obj) const throw ()
{
	if (this == obj)
		return true;

	if (obj)
	{
		const DSAPublicKey* pub = dynamic_cast<const DSAPublicKey*>(obj);
		if (pub)
		{
			if (pub->getParams().getP() != _params->getP())
				return false;

			if (pub->getParams().getQ() != _params->getQ())
				return false;

			if (pub->getParams().getG() != _params->getG())
				return false;

			if (pub->getY() != _y)
				return false;

			return true;
		}
	}

	return false;
}

const DSAParams& DSAPublicKeyImpl::getParams() const throw ()
{
	return *_params;
}

const BigInteger& DSAPublicKeyImpl::getY() const throw ()
{
	return _y;
}

const bytearray* DSAPublicKeyImpl::getEncoded() const throw ()
{
	if (!_enc)
	{
		try
		{
			ByteArrayOutputStream bos;
			BeeOutputStream bee(bos);

			bee.writeBigInteger(_params->getP());
			bee.writeBigInteger(_params->getQ());
			bee.writeBigInteger(_params->getG());
			bee.writeBigInteger(_y);
			bee.close();

			_enc = bos.toByteArray();
		}
		catch (IOException&)
		{
		}
	}

	return _enc;
}

const String& DSAPublicKeyImpl::getAlgorithm() const throw ()
{
	return ALGORITHM_DSA;
}

const String* DSAPublicKeyImpl::getFormat() const throw ()
{
	return &FORMAT_BEE;
}
